Vision Detection System Improvements

thomas0829 opened the pull request:
This PR enhances the rover's vision detection capabilities with improvements to AR detection and integration of a new object detection system.

## 1. AR Detection Enhancements

**Core Improvements:**
- Replace deprecated `estimatePoseSingleMarkers` API with `solvePnP` for better accuracy
- Enable AR detection support in simulator mode with proper configuration
- Fix CMakeLists.txt by removing non-existent test executables
- Improve initialization and logging for better debugging

**Runtime Control:**
- Add keyboard toggle control with `R` key (default: OFF)
- Clear old landmark data when disabling detection
- Background thread monitors enable/disable state
- State change logging (STARTED/STOPPED)

## 2. Object Detection Integration (NEW ✨)

**Model & Architecture:**
- Integrate OWL-ViT (Vision Transformer) model with LibTorch 2.1.0+
- Support for custom object classes (configurable)
- CUDA GPU acceleration for real-time performance
- Model file: `owlvit-cpp.pt` (excluded from git, 586MB)

**Features:**
- NMS (Non-Maximum Suppression) with IoU threshold 0.5 for duplicate removal
- Configurable confidence threshold (default: 0.6)
- Background detection thread with proper synchronization
- Integration with camera system (Mast Camera)
- Keyboard toggle control with `O` key (default: OFF)
- Automatic CUDA JIT optimization disabling (fixes nvrtc compilation errors)

**Testing & Visualization:**
- Side-by-side comparison test showing enabled/disabled states
- Bounding box visualization with confidence scores
- Detection count display
- Smart model path detection (multiple fallback locations)

**Code Structure:**
```
src/object-detection/
├── ObjectDetector.h/cpp       # Core detection engine with NMS
├── DetectionResult.h/cpp      # Detection result data structure
├── read_objects.h/cpp         # Integration interface
├── ObjectDetectionTest.cpp    # Visualization test
├── CMakeLists.txt             # Build configuration
└── generate_tokens.py         # Token generation utility
```

## 3. Unified Keyboard Control Interface

Both detection systems can now be controlled via keyboard in real-time:

| Key | Function | Default State |
|-----|----------|---------------|
| `O` | Toggle object detection | OFF |
| `R` | Toggle ArUco detection | OFF |
| `Q` | Quit program | - |

**Design Rationale:**
- Both systems disabled by default to reduce computational load
- Operators can enable only what they need during missions
- Clear console feedback for state changes
- Logging at INFO level for visibility

## 4. Optional LibTorch Support

**Flexibility:**
- Object detection is now **optional** - project builds without LibTorch
- CMake gracefully falls back if LibTorch not found
- Clear warning messages guide users to install LibTorch if desired
- README includes complete LibTorch installation instructions

**Installation:**
```bash
# Optional: Install LibTorch for object detection
export LIBTORCH_PATH=~/libtorch
cmake ../src
```

## Technical Details

**Dependencies Added:**
- LibTorch 2.1.0+ with CUDA support (optional)
- Environment variable: `LIBTORCH_PATH` for custom LibTorch location
- Alternative: `OWLVIT_MODEL_PATH` for custom model location

**Modified Files:**
- `src/Globals.h/cpp` - Added global enable flags
- `src/Rover.cpp` - Keyboard control loop with conditional compilation
- `src/ar/read_landmarks.cpp` - Enable/disable checking
- `src/network/MissionControlTasks.cpp` - ArUco detection guard
- `src/CMakeLists.txt` - Optional object detection module
- `src/object-detection/ObjectDetector.cpp` - CUDA JIT fix
- `src/object-detection/read_objects.cpp` - Smart model path detection
- `README.md` - LibTorch installation guide
- `.gitignore` - Ignore large model files

**CUDA Compatibility:**
- Automatic JIT optimization disabling prevents nvrtc errors
- No manual environment variable setup required
- Fusion strategy set to avoid `__ldg` undefined identifier issues

**Testing:**
```bash
# Build (object detection included if LibTorch found)
cd build
cmake ../src
make Rover -j4

# Run rover
./Rover -p arm

# Toggle object detection at runtime with 'O' key
# Toggle ArUco detection at runtime with 'R' key
```

**Without LibTorch:**
```bash
# Build succeeds, object detection disabled
cd build
cmake ../src  # Shows warning: "LibTorch not found. Object detection will be DISABLED."
make Rover -j4

# Run rover (ArUco detection still available)
./Rover -p arm

# Pressing 'O' shows: "Object detection not available (LibTorch not found)"
```

thomas0829 committed:
feat(ar): Replace deprecated estimatePoseSingleMarkers with solvePnP and add AR detection toggle

- Replace cv::aruco::estimatePoseSingleMarkers with cv::solvePnP for pose estimation
  to ensure compatibility with future OpenCV versions
- Add toggle functionality for AR detection:
  * Added setEnabled() and isEnabled() methods to Detector class
  * Added keyboard shortcut 'A' in ARTester to toggle detection on/off
  * Display keyboard controls help message on startup
- Add new camera configuration for Global Shutter Camera (ELP-USBFHD01M-L21)
  * Complete camera calibration with intrinsic parameters
  * Reprojection error: 0.103 pixels
- Update calibration.cpp to save format, framerate, and image dimensions
  required by Camera::getGSTPipe()

Related tasks:
- [x] Switch to solvePnP
- [x] Make AR detection toggle-able
- [x] Update calibration.cpp to save new camera calibration

thomas0829 committed:
feat(ar): Enable AR detection in simulator and add terminal output

- Uncommented intrinsic parameters in MastCameraCalibration.yml for simulator
- Initialize AR detection at startup after opening mast camera in Rover.cpp
- Load camera config in simulator_interface to provide intrinsic parameters
- Add AR marker detection in camera stream with terminal output
- Use std::map to deduplicate markers (show only first occurrence per ID)
- Add formatted printf output with marker ID, distance, and position
- Remove code comment markers per code review feedback
- Add simulator-specific AR test executables to CMakeLists.txt

Fixes simulator AR detection by ensuring intrinsic parameters are available
before AR detector initialization.

thomas0829 commented on src/Rover.cpp:
```diff
@@ -3,6 +3,7 @@
 #include "navtypes.h"
 #include "network/MissionControlProtocol.h"
 #include "world_interface/world_interface.h"
+#include "ar/read_landmarks.h"
```
💬 Added AR detection header for initialization at startup.

thomas0829 commented on src/Rover.cpp:
```diff
@@ -154,6 +155,18 @@ int main(int argc, char** argv) {
+		LOG_F(INFO, "AR landmark detection initialized successfully");
+	} else {
+		LOG_F(WARNING, "Failed to initialize AR landmark detection");
+	}
```
💬 Initialize AR detection at startup after opening mast camera. 
Opening camera first ensures intrinsic parameters are loaded before AR detector needs them.
This fixes the "Camera does not have intrinsic parameters" error in simulator.

thomas0829 commented on src/ar/ARTester.cpp:
```diff
@@ -130,6 +130,15 @@ int main(int argc, char* argv[]) {
+	std::cout << "  R - Show rejected points" << std::endl;
+	std::cout << "  L - Hide rejected points" << std::endl;
+	std::cout << "========================\n" << std::endl;
+
```
💬 Added keyboard controls help text including new 'A' key for toggling AR detection.

thomas0829 commented on src/camera/Camera.cpp:
```diff
@@ -192,4 +193,43 @@ void Camera::setDescription(std::string new_description) {
+bool Camera::isUndistortEnabled() const {
+	return *_undistort;
+}
+
```
💬 Added methods for undistortion control:
- initUndistortMaps(): Precomputes remap for efficient undistortion
- setUndistort(): Enable/disable automatic undistortion
- isUndistortEnabled(): Check current state

thomas0829 committed:
fix(ar): Remove non-existent test executables from CMakeLists.txt

Removed test_sim_camera and sim_ar_tester executables that don't have
corresponding source files, fixing CMake configuration error.

imisaacwu commented on src/network/MissionControlTasks.cpp:
```diff
@@ -130,6 +137,66 @@ void CameraStreamTask::task(std::unique_lock<std::mutex>&) {
 						const auto& encoder = stream_data.encoder;
 
+						// Detect and log AR markers if this is the mast camera
+						if (cam == Constants::MAST_CAMERA_ID && AR::isLandmarkDetectionInitialized()) {
```
Could y'all generalize this to have detection on any camera?

imisaacwu commented on src/network/MissionControlTasks.cpp:
```diff
@@ -130,6 +137,66 @@ void CameraStreamTask::task(std::unique_lock<std::mutex>&) {
+								std::vector<std::vector<cv::Point2f>> corners, rejectedPoints;
+								std::vector<int> ids;
+								cv::Ptr<cv::aruco::DetectorParameters> params = cv::aruco::DetectorParameters::create();
+								cv::aruco::detectMarkers(frame, AR::Markers::URC_MARKERS()->getDict(), 
```
Could you also generalize the marker type?

thomas0829 replied:
Created an aruco_marker_set variable at namespace scope initialized to URC_MARKERS(). Now to switch marker sets (e.g., for different competitions), we just need to change that one variable instead of modifying the detection code.

imisaacwu reviewed (changes requested):
Looks good, couple small things 👍

thomas0829 committed:
refactor(ar): Improve AR detection initialization and logging

- Remove camera opening from Rover initialization
- Load AR intrinsic params directly from config file
- Fix openCamera_() to load parameters from config
- Generalize AR detection to work on any camera
- Use reusable DetectorParameters
- Replace printf with LOG_F for consistent logging

thomas0829 committed:
Add object detection module with OWL-ViT integration

- Implemented ObjectDetector class with NMS and enable/disable functionality
- Added DetectionResult data structure for detection output
- Integrated object detection into Rover with keyboard control ('O' key)
- Added read_objects interface for background detection loop
- Created visualization test with side-by-side comparison
- Added Globals::objectDetectionEnabled flag for runtime control
- Updated CMakeLists.txt to include LibTorch dependencies
- Ignore large model files (.pt) and test images in gitignore

thomas0829 committed:
Add keyboard control for ArUco detection toggle

- Added Globals::arucoDetectionEnabled flag (default: false)
- Added 'R' key to toggle ArUco detection on/off
- Modified detectLandmarksLoop() to check enabled flag
- Clear old landmarks when disabling detection
- Modified MissionControlTasks to check enabled flag before detection
- Added state change logging (STARTED/STOPPED)
- readLandmarks() returns empty data when disabled

thomas0829 committed:
Fix duplicate function definitions after rebase

thomas0829 committed:
Replace hardcoded LibTorch path with environment variable

- Use LIBTORCH_PATH environment variable instead of absolute path
- Falls back to CMAKE_PREFIX_PATH or system installation
- Usage: export LIBTORCH_PATH=/path/to/libtorch or cmake -DCMAKE_PREFIX_PATH=/path/to/libtorch

thomas0829 committed:
Make object detection optional when LibTorch not available

- Changed find_package(Torch) from REQUIRED to QUIET
- Added ENABLE_OBJECT_DETECTION compile definition when Torch is found
- Added conditional compilation guards (#ifdef) in Rover.cpp
- Object detection features gracefully disabled without LibTorch
- 'O' keyboard command shows warning when feature unavailable
- Clear CMake warnings guide users to install LibTorch if needed

thomas0829 committed:
Add LibTorch installation instructions to README

- Document object detection as an optional feature
- Provide step-by-step LibTorch installation guide
- Explain how to set LIBTORCH_PATH environment variable
- Clarify that build continues without LibTorch (feature disabled)

thomas0829 committed:
Fix CUDA JIT compilation issues and improve model path handling

- Disable PyTorch JIT optimizations and fusion to prevent nvrtc errors
- Add automatic model path detection (checks multiple locations)
- Support OWLVIT_MODEL_PATH environment variable for custom paths
- Remove need for PYTORCH_JIT=0 environment variable
- Update README to reflect simplified setup

thomas0829 committed:
Add automatic OWL-ViT model download functionality

- Add ModelDownloader module for automatic model download from Hugging Face
- Integrate auto-download into object detection initialization
- Model is automatically downloaded if not found locally
- Uses wget/curl for pure C++ implementation
- Support OWLVIT_MODEL_URL environment variable for custom URLs
- Model hosted at thomas0829/OWL-ViT on Hugging Face

thomas0829 committed:
Add RealSense camera support and additional object detection features

- Add RealSenseCamera class for Intel RealSense depth cameras
- Add camera calibration tool (CalibrateCamera.cpp)
- Add RealSense test program
- Update object detection with improved distance estimation
- Add calibrate_and_test.sh script
- Update world interface for object detection integration

thomas0829 committed:
Add model fine-tuning pipeline and improve object detection

- Add OWL-ViT fine-tuning scripts (train, export, dataset, capture)
- Update ObjectDetector to use CLIP normalization and vision-only model
- Export model with baked text embeddings (no text encoder at runtime)
- Lower default confidence threshold to 0.75
- Add __pycache__ and dataset artifacts to .gitignore

thomas0829 committed:
Add add_real_data.py script and update .gitignore

thomas0829 committed:
Update README with complete workflow documentation

thomas0829 committed:
Expand README with detailed setup and workflow steps

thomas0829 committed:
Add web image download workflow to README

thomas0829 committed:
Replace v7 placeholder with my_run in README

thomas0829 committed:
Add static image testing option to README
